detox/src/client/actions/SyncStatusFormatter.js

```diff
@@ -0,0 +1,277 @@
+const Ajv = require('ajv');
+
+const DetoxRuntimeError = require('../../errors/DetoxRuntimeError');
+
+class SyncStatusFormatter {
+  static formatJSONStatus(jsonStatus) {
+    if (!SyncStatusFormatter.isValidJSONStatus(jsonStatus)) {
+      throw new DetoxRuntimeError(`Given sync status is not compatible with the status schema, ` +
+        `given status: ${JSON.stringify(jsonStatus)}`);
+    }
+
+    if (SyncStatusFormatter.isAppIdle(jsonStatus)) {
+      return 'The app seems to be idle';
+    }
+
+    if (jsonStatus.busy_resources === undefined) {
+      throw new DetoxRuntimeError(`Given sync status is invalid, app is busy but busy-resources are not defined`);
+    }
+    const resourcesDescriptions = SyncStatusFormatter.resourcesDescriptionsFromJSON(jsonStatus.busy_resources);
+    return `The app is busy with the following tasks:\n${resourcesDescriptions.join('\n')}`;
+  }
+
+  static isValidJSONStatus(jsonStatus) {
+    const ajv = new Ajv();
+    const isValidStatus = ajv.compile(SyncStatusFormatter.jsonStatusSchema());
+    return isValidStatus(jsonStatus);
+  }
+
+  static jsonStatusSchema() {
+    return {
+      type: 'object',
+      properties: {
+        app_status: { 'enum': ['busy', 'idle'] },
+        busy_resources: {
+          type: 'array',
+          items: {
+            type: 'object',
+            properties: {
+              name: { 'type': 'string' },
+              description: {
+                type: 'object',
+                additionalProperties: true
+              }
+            },
+            required: ['name'],
+            additionalProperties: false
+          }
+        }
+      },
+      required: ['app_status'],
+      additionalProperties: false
+    };
+  }
+
+  static isAppIdle(jsonStatus) {
+    return jsonStatus.app_status === 'idle' ||
+      (jsonStatus.busy_resources !== undefined && !jsonStatus.busy_resources.length);
+  }
+
+  static resourcesDescriptionsFromJSON(jsonDescriptions) {
+    let descriptions = [];
+    for (const jsonDescription of jsonDescriptions) {
+      const description = SyncStatusFormatter.resourceDescriptionFromJSON(jsonDescription);
+      descriptions.push(description);
+    }
+
+    return descriptions;
+  }
+
+  static resourceDescriptionFromJSON(jsonDescription) {
+    const resourceName = jsonDescription.name;
+    const properties = jsonDescription.description;
+
+    switch (resourceName) {
+      case 'delayed_perform_selector':
+        return SyncStatusFormatter.delayedPerformSelectorResourceDescriptionFromProperties(properties);
```

Comment:
Wouldn't it be easier to have a map?

```js
const formatter = this._formatters[resourceName];
if (!formatter) {
  throw ...
}
return formatter(properties);
```

where:

```js
this._formatters = {
  delayed_perform_selector: () => {...},
  dispatch_queue: () => {...}
};
```

Reply:
both easy, but I agree that map will look better here :)
done